src/dbnode/storage/namespace.go

```diff
@@ -1602,3 +1604,171 @@ func (n *dbNamespace) FlushState(shardID uint32, blockStart time.Time) (fileOpSt
+	nsCtx namespace.Context,
+	shard databaseShard,
+	pm persist.Manager,
+	resources coldFlushReuseableResources,
+	multiErr xerrors.MultiError,
+) xerrors.MultiError {
+	// NB(rartoul): This value can be used for emitting metrics, but should not be used
+	// for business logic.
+	callStart := n.nowFn()
+
+	// NB(bodu): The in-mem index will lag behind the TSDB in terms of new series writes. For a period of
+	// time between when we rotate out the active cold mutable index segments (happens here) and when
+	// we actually cold flush the data to disk we will be making writes to the newly active mutable seg.
+	// This means that some series can live doubly in-mem and loaded from disk until the next cold flush
+	// where they will be evicted from the in-mem index.
+	var (
+		onColdFlushDone OnColdFlushDone
+		err             error
+	)
+	if n.reverseIndex != nil {
+		onColdFlushDone, err = n.reverseIndex.ColdFlush([]databaseShard{shard})
+		if err != nil {
+			n.metrics.writeAggregated.ReportError(n.nowFn().Sub(callStart))
+			return multiErr.Add(
+				fmt.Errorf("error preparing to coldflush a reverse index for shard %d: %v",
+					shard.ID(),
+					err))
+		}
+	}
+
+	onColdFlushNs, err := n.opts.OnColdFlush().ColdFlushNamespace(n)
+	if err != nil {
+		n.metrics.writeAggregated.ReportError(n.nowFn().Sub(callStart))
+		return multiErr.Add(
+			fmt.Errorf("error preparing to coldflush a namespace for shard %d: %v",
+				shard.ID(),
+				err))
+	}
+
+	flushPersist, err := pm.StartFlushPersist()
+	if err != nil {
+		n.metrics.writeAggregated.ReportError(n.nowFn().Sub(callStart))
+		return multiErr.Add(
+			fmt.Errorf("error starting flush persist for shard %d: %v",
+				shard.ID(),
+				err))
+	}
+
+	localErrors := xerrors.NewMultiError()
+	shardColdFlush, err := shard.ColdFlush(flushPersist, resources, nsCtx, onColdFlushNs)
+	if err != nil {
+		detailedErr := fmt.Errorf("shard %d failed to compact: %v", shard.ID(), err)
+		localErrors = localErrors.Add(detailedErr)
+	}
+
+	// We go through this error checking process to allow for partially successful flushes.
+	indexColdFlushError := onColdFlushNs.Done()
+	if indexColdFlushError == nil && onColdFlushDone != nil {
+		// Only evict rotated cold mutable index segments if the index cold flush was successful
+		// or we will lose queryability of data that's still in mem.
+		indexColdFlushError = onColdFlushDone()
+	}
+	if indexColdFlushError == nil {
+		// NB(bodu): We only want to complete data cold flushes if the index cold flush
+		// is successful. If index cold flush is successful, we want to attempt writing
+		// of checkpoint files to complete the cold data flush lifecycle for successful shards.
+		localErrors = localErrors.Add(shardColdFlush.Done())
+	}
+	localErrors = localErrors.Add(indexColdFlushError)
+	err = flushPersist.DoneFlush()
+	localErrors = multiErr.Add(err)
+
+	res := localErrors.FinalError()
+	n.metrics.writeAggregated.ReportSuccessOrError(res, n.nowFn().Sub(callStart))
```

Comment:
Remove all these since they duplicate the top level n.metrics.writeAggregated.ReportErrorOrSuccess(...) call from `AggregateTiles(...)`

Reply:
Same as above.